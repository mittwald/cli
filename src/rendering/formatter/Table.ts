import { getTerminalWidth } from "../lib/getTerminalWidth.js";

export type ListColumn<TItem> = {
    header?: string;
    extended?: boolean;
    get?: (row: TItem) => string | undefined;
    minWidth?: number;
};

export type ListColumns<TItem> = {
    [key: string]: ListColumn<TItem>;
};

export interface TableOptions {
    extended: boolean;
    truncate: boolean;
}

export default class Table<TItem> {
    private columns: ListColumns<TItem>;
    private opts: TableOptions;

    public constructor(columns: ListColumns<TItem>, opts: Partial<TableOptions>) {
        this.columns = columns;
        this.opts = {
            extended: false,
            truncate: true,
            ...opts,
        }
    }

    public render(data: TItem[]): string {
        const availableWidth = getTerminalWidth();
        const renderedItems: string[][] = [];

        const colList = Object.entries(this.columns).filter(([, spec]) => this.opts.extended ? true : !spec.extended);

        const reservedWidths = colList.map(([, spec]) => spec.minWidth || 0);
        const dynamicWidths = colList.map(() => 0);
        const reservedForColumnGaps = (colList.length - 1) * 2;

        for (let idx = 0; idx < data.length; idx ++) {
            const rendered = colList.map(([key, spec]) => {
                if (spec.get) {
                    return spec.get(data[idx]);
                }
                return (data[idx] as any)[key];
            });

            rendered.forEach((val, idx) => {
                dynamicWidths[idx] = Math.max(dynamicWidths[idx], colList[idx][1].minWidth ? 0 : val.length);
            });

            renderedItems.push(rendered);
        }

        const requiredWidth = dynamicWidths.reduce((prev, current) => prev + current, 0);
        const reservedWidthSum = reservedWidths.reduce((prev, current) => prev + current, 0);

        const definiteColWidths = colList.map(([, spec], idx) => {
            if (spec.minWidth) {
                return spec.minWidth;
            }

            if (availableWidth) {
                const allocatable = availableWidth - reservedWidthSum - reservedForColumnGaps;
                const fractionalWidth = dynamicWidths[idx] / requiredWidth;
                return Math.floor(allocatable * fractionalWidth);
            }

            return dynamicWidths[idx];
        });
        
        let output = "";

        const headerColumns = colList.map(([key, value]) => (value.header ?? key).toUpperCase());
        const truncatedHeaderColumns = headerColumns.map((val, idx) => val.substring(0, definiteColWidths[idx]));
        const paddedHeaderColumns = truncatedHeaderColumns.map((val, idx) => val + " ".repeat(definiteColWidths[idx] - val.length));

        output += paddedHeaderColumns.join("  ") + "\n";
        output += paddedHeaderColumns.map(v => "".repeat(v.length)).join("  ") + "\n";

        for (let idx = 0; idx < renderedItems.length; idx ++) {
            const truncatedRows = renderedItems[idx].map((val, colIdx) => val.substring(0, definiteColWidths[colIdx]));
            const paddedRows = truncatedRows.map((val, colIdx) => val + " ".repeat(definiteColWidths[colIdx] - val.length));

            output += paddedRows.join("  ") + "\n";
        }

        return output;
    }
}