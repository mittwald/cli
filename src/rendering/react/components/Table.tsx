import React, { ReactElement } from "react";
import { Box, Text } from "ink";
import { useRenderContext } from "../context.js";
import { RenderJson } from "../json/RenderJson.js";

interface Props<TData> {
    data: TData[];
    columns?: string[];
}

// TODO: Dynamic columns and typed data
export function Table<TData>(props: Props<any>): ReactElement {
    const { columns, data } = props;

    const { renderAsJson } = useRenderContext();

    if (renderAsJson) {
        return <RenderJson name="data" data={data} />;
    }

    return (
        <Box flexDirection="column" margin={1} gap={0} padding={0}>
            <Box margin={0} gap={1} padding={0}>
                <Box
                    borderStyle="single"
                    minWidth={38}
                    padding={0}
                    margin={0}
                    borderBottom
                    borderTop={false}
                    borderLeft={false}
                    borderRight={false}
                >
                    <Text bold>ID</Text>
                </Box>

                <Box
                    borderStyle="single"
                    width="50%"
                    padding={0}
                    margin={0}
                    borderBottom
                    borderTop={false}
                    borderLeft={false}
                    borderRight={false}
                >
                    <Text bold>Name</Text>
                </Box>

                <Box
                    width="40%"
                    borderStyle="single"
                    padding={0}
                    margin={0}
                    borderBottom
                    borderTop={false}
                    borderLeft={false}
                    borderRight={false}
                >
                    <Text bold>Status</Text>
                </Box>
            </Box>
            <Box flexDirection="column">
                {props.data.map((p) => (
                    <Box key={p.id} gap={1}>
                        <Box minWidth={38} padding={0}>
                            <Text>{p.id}</Text>
                        </Box>

                        <Box width="50%" padding={0}>
                            <Text>{p.description}</Text>
                        </Box>

                        <Box width="40%" padding={0}>
                            <Text>{p.readiness}</Text>
                        </Box>
                    </Box>
                ))}
            </Box>
            <Box marginTop={5}>
                <Text dimColor>Columns: {JSON.stringify(props.columns)}</Text>
            </Box>
        </Box>
    );
}
