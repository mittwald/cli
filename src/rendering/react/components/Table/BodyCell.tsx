import * as Model from "./model/index.js";
import React, { FC } from "react";
import { CellContainer } from "./CellContainer.js";
import { CellContent } from "./CellContent.js";

interface Props {
  cell: Model.Cell<unknown>;
}

export const BodyCell: FC<Props> = (props) => {
  const { cell } = props;

  return (
    <CellContainer col={cell.column}>
      <CellContent data={cell.data} />
    </CellContainer>
  );
};
