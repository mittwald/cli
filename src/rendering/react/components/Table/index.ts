export { Table } from "./Table.js";
