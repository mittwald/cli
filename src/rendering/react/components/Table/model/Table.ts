import { Row } from "./Row.js";
import { Column } from "./Column.js";
import { ObservableValue } from "../../../lib/observable-value/ObservableValue.js";
import { ColumnOptionsInput } from "./ColumnOptions.js";
import { TableRenderSetupOutput } from "../../../../setup/TableRenderSetup.js";
import { ColumnName } from "./ColumnName.js";

export type TableOptions = TableRenderSetupOutput;
export type ColumnOptionsInputMap = Record<string, ColumnOptionsInput>;

export class Table<T> {
  public readonly rows: Row<T>[];
  public readonly columns: Column[];
  public readonly overallWidth = new ObservableValue(0);
  public readonly tableOptions?: TableOptions;

  public constructor(
    data: T[],
    tableOptions?: TableOptions,
    columnOptionsInput?: ColumnOptionsInputMap,
  ) {
    this.tableOptions = Object.freeze(tableOptions);
    this.rows = this.buildRows(data);
    this.columns = this.buildColumns(this.rows, columnOptionsInput);
  }

  private buildRows<T>(data: T[]): Row<T>[] {
    return data.map((d) => Row.fromObject<T>(this, d));
  }

  private handleColumnWidthChanged(): void {
    const overallWidth = this.columns.reduce(
      (width, col) => width + col.maxCellWidth.value,
      0,
    );
    this.overallWidth.updateValue(overallWidth);
  }

  private buildColumns<T>(
    rows: Row<T>[],
    columnOptions?: ColumnOptionsInputMap,
  ): Column[] {
    const columnsMap = new Map<string, Column>();

    const addColumn = (columnName: string): void => {
      const options = columnOptions?.[columnName];
      const column = new Column(this, columnName, options);
      columnsMap.set(columnName, column);
      column.maxCellWidth.observe(() => this.handleColumnWidthChanged());
    };

    for (const row of rows) {
      const names = columnOptions
        ? Object.entries(columnOptions).map(([name]) => name)
        : row.collectColumnNamesFromCells().map((n) => n.value);

      for (const name of names) {
        if (!columnsMap.has(name)) {
          addColumn(name);
        }
      }
    }

    return Array.from(columnsMap.values());
  }

  public getColumn(name: ColumnName | string): Column {
    const col = this.columns.find((col) => col.name.matches(name));

    if (!col) {
      throw new Error(`Could not get column with ID ${name}`);
    }

    return col;
  }
}
