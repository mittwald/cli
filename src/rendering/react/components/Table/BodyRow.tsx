import React, { ReactElement } from "react";
import * as Model from "./model/index.js";
import { BodyCell } from "./BodyCell.js";
import { RowContainer } from "./RowContainer.js";

interface Props<TData> {
  row: Model.Row<TData>;
}

export function BodyRow<TData>(props: Props<TData>): ReactElement {
  const { row } = props;

  return (
    <RowContainer>
      {row.table.columns.map((col) => (
        <BodyCell cell={row.getCell(col.name)} key={col.name.value} />
      ))}
    </RowContainer>
  );
}
