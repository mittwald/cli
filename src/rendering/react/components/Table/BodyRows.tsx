import React, { FC } from "react";
import { BodyRow } from "./BodyRow.js";
import * as Model from "./model/index.js";

interface Props {
  table: Model.Table;
}

export const BodyRows: FC<Props> = (props) => {
  const { table } = props;

  return (
    <>
      {table.rows.map((row) => (
        <BodyRow row={row} key={row.index} />
      ))}
    </>
  );
};
