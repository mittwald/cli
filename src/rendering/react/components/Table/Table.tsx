import * as Model from "./model/index.js";
import { TableColumnsInput } from "./model/index.js";
import React, { ReactNode, useRef } from "react";
import { Box } from "ink";
import { useRenderContext } from "../../context.js";
import { RenderJson } from "../../json/RenderJson.js";
import { MeasureContextProvider } from "../../measure/MeasureContextProvider.js";
import { HeaderRow } from "./HeaderRow.js";
import { BodyRows } from "./BodyRows.js";
import { TableRenderSetupOutput } from "../../../setup/TableRenderSetup.js";
import { buildTableContext, TableContextProvider } from "./context.js";

interface Props<TData> {
  data: TData[];
  setup: TableRenderSetupOutput;
  columns?: TableColumnsInput<TData>;
}

export function Table<TData>(props: Props<TData>): ReactNode {
  const { data, columns, setup } = props;

  const { renderAsJson } = useRenderContext();

  if (renderAsJson) {
    return <RenderJson name="data" data={data} />;
  }

  const table = useRef(new Model.Table(data, setup, columns)).current;

  return (
    <TableContextProvider
      value={buildTableContext({
        setup,
      })}
    >
      <MeasureContextProvider>
        <Box flexDirection="column">
          <HeaderRow marginBottom={1} table={table} />
          <BodyRows table={table} />
        </Box>
      </MeasureContextProvider>
    </TableContextProvider>
  );
}
