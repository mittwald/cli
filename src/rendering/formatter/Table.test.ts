import Table from "./Table.js";
import { expect } from "@jest/globals";

function effectiveStringLength(str: string) {
  const backspaceCount = (str.match(/[\b]/g) ?? []).length;
  return str.length - backspaceCount * 2;
}

describe("Table", () => {
  it("should spread a single column across the max width", () => {
    const t = new Table(
      {
        a: {
          header: "A",
        },
      },
      { maxWidth: 20, chalkOptions: { level: 0 } },
    );

    const rendered = t.render([{ a: "Foo" }, { a: "Bar" }]);

    expect(rendered).toMatchSnapshot();
    expect(rendered.split("\n")[0].length).toBe(20);
  });

  it("should truncate too long values using backspaces", () => {
    const t = new Table(
      {
        a: {
          header: "A",
        },
      },
      { maxWidth: 20, chalkOptions: { level: 0 } },
    );

    const rendered = t.render([
      { a: "Foo".repeat(10) },
      { a: "Bar".repeat(10) },
    ]);

    const lines = rendered.split("\n");

    expect(rendered).toMatchSnapshot();
    expect(lines[0].length).toBe(20);
    expect(effectiveStringLength(lines[2])).toBe(20);
  });
});
