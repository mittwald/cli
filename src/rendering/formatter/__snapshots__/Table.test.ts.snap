// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table should give expanding columns the remaining space 1`] = `
"A    B           C  
───  ──────────  ───
aaa  bbbbbbbbb…  ccc
aaa  bbbbbbbbb…  ccc
"
`;

exports[`Table should not truncate too long values when truncation is disabled 1`] = `
"A                                                           
────────────────────────────────────────────────────────────
FooFooFooFooFooFooFooFooFooFooFooFooFooFooFooFooFooFooFooFoo
BarBarBarBarBarBarBarBarBarBarBarBarBarBarBarBarBarBarBarBar
"
`;

exports[`Table should respect the minimum column width 1`] = `
"A                B 
───────────────  ──
a                b…
a                b…
"
`;

exports[`Table should spread a single column across the max width 1`] = `
"A                   
────────────────────
Foo                 
Bar                 
"
`;

exports[`Table should still give non-expanding columns a bit more space if available 1`] = `
"A    B                               C  
───  ──────────────────────────────  ───
aaa  bbbbbbbbbbbbbbbbbbbbbbbbbbbbb…  ccc
aaa  bbbbbbbbbbbbbbbbbbbbbbbbbbbbb…  ccc
"
`;

exports[`Table should truncate too long values 1`] = `
"A                   
────────────────────
FooFooFooFooFooFooF…
BarBarBarBarBarBarB…
"
`;

exports[`Table without max width should give each column the maximum content width when no max width  1`] = `
"A    B                                         C  
───  ────────────────────────────────────────  ───
aaa  bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb  ccc
aaa  bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb  ccc
"
`;
